test/runner/retry_test.dart

```diff
@@ -0,0 +1,148 @@
+    test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+    test.shouldExit(0);
+  });
+
+  group("retries tests", () {
+    test("and eventually passes for valid tests", () {
+      d
+          .file(
+              "test.dart",
+              """
+              import 'dart:async';
+
+              import 'package:test/test.dart';
+
+              int attempt = 0;
+              void main() {
+                test("eventually passes", () {
+                 attempt++;
+                 if(attempt <= 2) {
+                   throw new TestFailure("oh no");
+                 }
+                }, retry: 2);
+              }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+      test.shouldExit(0);
+    });
+
+    test("and ignores previous errors", () {
+      d
+          .file(
+              "test.dart",
+              """
+              import 'dart:async';
+
+              import 'package:test/test.dart';
+
+              int attempt = 0;
+              Completer completer = new Completer();
+              void main() {
+                test("failure", () {
+                  attempt++;
+                  if (attempt == 1) {
+                    completer.future.then((_) => throw 'some error');
+                    throw new TestFailure("oh no");
+                  }
+                  completer.complete(null);
+                }, retry: 1);
+              }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+      test.shouldExit(0);
+    });
+
+    test("and eventually fails for invalid tests", () {
+      d
+          .file(
+              "test.dart",
+              """
+              import 'dart:async';
+
+              import 'package:test/test.dart';
+
+              void main() {
+                test("failure", () {
+                 throw new TestFailure("oh no");
+                }, retry: 2);
+              }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("-1: Some tests failed.")));
+      test.shouldExit(1);
+    });
+
+    test("only after a failure", () {
+      d
+          .file(
+              "test.dart",
+              """
+              import 'dart:async';
+
+              import 'package:test/test.dart';
+
+              int attempt = 0;
+              void main() {
+                test("eventually passes", () {
+                attempt++;
+                if (attempt != 2){
+                 throw new TestFailure("oh no");
+                }
+                }, retry: 5);
+          }
+          """)
+          .create();
+
+      var test = runTest(["test.dart"]);
+      test.stdout.expect(consumeThrough(contains("+1: All tests passed!")));
+      test.shouldExit(0);
+    });
+  });
```

Comment:
We should also test what happens if a retry test fails after it reports success. Something like

```dart
void main() {
  var completer1 = new Completer();
  var completer2 = new Completer();
  test("first", () {
    completer1.future.then((_) {
      completer2.complete();
      throw "oh no";
    });
  });

  test("second", () async {
    completer1.complete();
    await completer2.future;
  });
}
```